src/Dofs/sparsity_pattern.jl

```diff
@@ -1,83 +1,473 @@
+    # interface_coupling, topology)
+end
 
-Create the sparsity pattern corresponding to the degree of freedom
-numbering in the [`DofHandler`](@ref). Return a `SparseMatrixCSC`
-with stored values in the correct places.
 
-The keyword arguments `coupling` and `cross_coupling` can be used to specify how fields (or components) in the dof
-handler couple to each other. `coupling` and `cross_coupling` should be square matrices of booleans with
-number of rows/columns equal to the total number of fields, or total number of components,
-in the DofHandler with `true` if fields are coupled and `false` if
-not. By default full coupling is assumed inside the element with no coupling between elements.
+# # Arguments
+#  - `sp`: the sparsity pattern to add entries to
+#  - `dh`: the DofHandler for which to generate the pattern
+#  - `ch`: the ConstraintHandler. Used to add the non-zero entries resulting from
+#    `AffineConstraint`s (and `PeriodicDirichlet` constraints, which uses `AffineConstraint`s
+#    internally). See [`add_constraint_entries!`](@ref).
+# # Keyword arguments
+#  - `keep_constrained`: whether or not entries for constrained DoFs should be kept
+#    (`keep_constrained = true`) or eliminated (`keep_constrained = false`) from the sparsity
+#    pattern. `keep_constrained = false` requires passing the ConstraintHandler.
+#  - `coupling`: the coupling between fields/components within each cell. By default
+#    (`coupling = nothing`) it is assumed that all DoFs in each cell couple with each other.
 
-If `topology` and `cross_coupling` are passed, dof of fields with discontinuous interpolations are coupled between elements according to `cross_coupling`.
+"""
+    add_cell_entries!(
+        sp::AbstractSparsityPattern,
+        dh::DofHandler,
+        ch::Union{ConstraintHandler, Nothing} = nothing;
+        keep_constrained::Bool = true,
+        coupling::Union{AbstractMatrix{Bool}, Nothing}, = nothing
+    )
+
+Add entries to the sparsity pattern `sp` corresponding to DoF couplings within the cells as
+described by the DofHandler `dh`.
+
+# Keyword arguments
+ - `keep_constrained`: whether or not entries for constrained DoFs should be kept
+   (`keep_constrained = true`) or eliminated (`keep_constrained = false`) from the sparsity
+   pattern. `keep_constrained = false` requires passing the ConstraintHandler `ch`.
+ - `coupling`: the coupling between fields/components within each cell. By default
+   (`coupling = nothing`) it is assumed that all DoFs in each cell couple with each other.
+"""
+function add_cell_entries!(
+        sp::AbstractSparsityPattern,
+        dh::DofHandler, ch::Union{ConstraintHandler, Nothing} = nothing;
+        keep_constrained::Bool = true, coupling::Union{AbstractMatrix{Bool}, Nothing} = nothing,
+    )
+    # Expand coupling from nfields × nfields to ndofs_per_cell × ndofs_per_cell
+    # TODO: Perhaps this can be done in the loop over SubDofHandlers instead.
```

Comment:
The resulting pattern should be equivalent in this case anyway, right?

Reply:
Yes, just a bit unnecessary to expand them into a collection instead of on-demand.